### DISCUSSION TITLE
RFC: UI Layout Changes (Part 1)

### DISCUSSION THREAD

**Discussion by n0toose** [tenacityteam/tenacity-legacy#398] [Show and tell]
Hi, I spent some time today experimenting with the UI and I explored a couple of different ideas.

![2021-07-27_09-23](https://user-images.githubusercontent.com/30193966/127158311-faf12b39-c8ab-437b-808d-12bcb930a06e.png)

![2021-07-27_09-26](https://user-images.githubusercontent.com/30193966/127158350-2594c035-72c8-4ed5-a171-67ee8429423d.png)

The primary idea here is to move the volume indicators on a new bar, while taking advantage of the vast amount of space that's available to us. This obviously isn't final, just a general idea, but it will definitely need a lot of polishing if I were to actually implement something like this (since I only really moved widgets around). I'm sort of skeptical about the concept of sacrificing horizontal screen space in order to make the volume bars more easily readable. I'd like to hear your feedback.

Some ideas haven't been included in the screenshot, as the screenshot depict an actual instance of Tenacity running on top of the master branch (a21a554f5bf366d9286556273d27db5eba63b99e). They are out of scope for the time being, until I (or someone else) come up with some tangible examples. Some of them include
- breaking the bar that contains the "Undo/Redo" icons, the "Zoom In/Out", etc. into more bars for better customization.
- changing the order of the widget with the play/pause/record/rewing/fast-forward buttons
- making some objects more customizable.
- somehow doing away with Audacity's tendency of aligning everything towards the left.
- making settings easier to use.

**Comment by akleja** [comment]
I love this! The transport looks way more usable.
I would place the output and input faders next to the meters as well.

I've been writing a feature request https://github.com/tenacityteam/tenacity/issues/399 today for changes to the meters as well. There I propose docking areas on both sides of the window to make vertical meters useful.

I've also been thinking about ways to improve the device/channel ui, It's a bit confusing right now imo. It might be a good idea to indicate which is protocol/driver, and which is device. And the channel-count label "recording channels" could easily be interpreted not as number of channels to record, but channel routing, and it's also strange to me to have the label included in the drop down of the first two options.

**Reply by n0toose** [reply]
Would you like to get in a call and work on that sort of stuff together at some point? (We'd love to have you in our discussion rooms, BTW!)

**Reply by akleja** [reply]
Yeah absolutely, that would be nice!
Yeah sure where do I find these discussion rooms?

**Reply by Semisol** [reply]
@akleja
| Purpose | Libera | Matrix | Is bridged |
|-|-|-|-|
| Matrix, recommended for voice | N/A | https://matrix.to/#/%23tenacity:semisol.dev | No |
| General discussion | #tenacity | https://matrix.to/#/%23tenacity:libera.chat | Yes |
| Offtopic | #tenacity-watercooler | https://matrix.to/#/%23tenacity-dev:libera.chat | Yes |
| Dev | #tenacity-dev | https://matrix.to/#/%23tenacity-dev:libera.chat | Yes |

**Reply by akleja** [reply]
@Semisol Oh thanks! Didn't understand if @panos referred to the irc channels or something else.

**Reply by akleja** [reply]
I'll set up a matrix account in a while so I can join voice chats

**Reply by Semisol** [reply]
@akleja Or we can give the Jitsi URL, but I think people who are muted will prefer to talk in the Matrix chat instead of Jitsi.

**Reply by akleja** [reply]
@Semisol It'll be fine with matrix later on, but for now I'll stick with IRC / emails because I'm not really used to using voice chat as it always makes me nervous/stressed out.

**Comment by akleja** [comment]
I've been looking into the toolbar dock code a bit, and begun experimenting with adding vertical docks.
There's still a long road ahead and lots of work to be done yet regarding sizing/moving/flowing of toolbars, and getting to the bottom of some strange behaviors. So consider this screenshot very much an early wip.
![meterdock](https://user-images.githubusercontent.com/49785979/127780436-d3da317d-5d11-4f8a-bdbb-beb9a068e426.png)

**Reply by poyribn** [reply]
This improvement is worth pursuing.

The pre-fork UI feels like it was made for 4:3 Aspect Ratio Screens.

Vertical Toolbars like this will enable customized UI layouts that use the space on today's 16:9 (and wider) screens much better.
If you use a vertical OS taskbar, you know what I mean.

**Reply by akleja** [reply]
Cool, I'll keep working on it! There are quite som things to figure out still though manly concerning sizing. Currently working on alternative layouts for toolbars docked on vertical toolbars, which makes it possible to place the long edit toolbars on the side as well, and makes the toolbar handles appear horizontally.

Yeah good point. Also I'm way too used to reading vertical meters from using analog mixers quite a bit.

**Reply by n0toose** [reply]
I am starting to think that moving the volume/microphone volume bars from the top bar to the bottom bar and moving the microphone bar like in the latter screenshot may be a better, less ground-shifting and more ergonomic idea. Not sure.

**Comment by akleja** [comment]
Some more updates on my experiments:
![1627954726](https://user-images.githubusercontent.com/49785979/127944409-0a6d338c-df03-4bb4-8a43-9b86bfd0c6a9.png)
* Vertical docks now on both sides
* Ruler moved in between vertical dock
* It is now possible to dock in multiple columns in the vertical dock, even though the behavior needs some polish
* Toolbars flip vertically when docked in vertical docks
* Removal of some bevels and lines that were disruptive, and changing the look of the toolbar handles

To do:
* Flipped slider toolbars
* Automatic sizing of meter height
* Flip floating toolbars as well on resize
* Break some toolbars down into multiple
* Prevent unflippable toolbars from docking in vertical toolbars
* Possibly changes to toolbar placement
* Optional small transport (maybe) or make it resizeable.
* Save positions of vertically docked toolbars
* Ensure tab order works correctly
* Code cleanup (a lot)

I will try to have a PR up for testing in the near future. Just have to go through some key points of my to-do list, and split up changes into appropriate commits.

So far the classic layout is still possible, it's just a matter of how you choose to position the docks.

**Reply by akleja** [reply]
Another thing I just realized is that it should be possible to write a custom drop down control instead of relying on system ones, much alike NumericTextCtrl, but that's it's own project I guess.

**Reply by n0toose** [reply]
I (and others that I asked on Mastodon) seem to have a concern regarding the positioning of the tools. I think it makes things harder to reach, rather then easier.

I am also not sure whether the sample rate, which is an option that nobody really touches, should be that close to the 5 buttons. I agree with keeping the Snap-To and the options revolving around selecting the samples together. We also probably have to do something about the Undo/Redo buttons.

**Reply by akleja** [reply]
I have to clarify that this isn't in any way a final layout or a proposal of a layout, rather a screenshot of what's possible with vertical dockers in my current wip build. Since you can move everything around the layout is up to the user. What's a good default layout is however a discussion to be had. 

I really prefer the the tools on the side, and vertical dockers gives that option. No one has to actually position them that way. I just like the extra flexibility of having that option. I always felt the huge number of top toolbars makes the interface both confusing and ugly. For me having the tools on the side makes the tools way easier to find and select, but maybe that's just me. 

Yeah I really agree that the project sample rate should ideally not even be a toolbar. perhaps there should be a project settings section in the prefs window?

I haven't started breaking apart toolbars yet. I agree with breaking undo/redo and zoom controls into separate toolbars. but it's on the to do list.

As for the snapping/selection toolbar that's it's own mess.
I think it's a bit unclear from a ui perspective what that snap setting does. It's snap cursor and selection, but that's not super obvious.
I also find it strange that the time division from the selection toolbar isn't shown in the ruler. If I choose, say pal frames, it would make sense to see it visually represented. If this is done, i think the snap setting could be reduced to on/off instead of having that prior/nearest option. 
It's even more baffling that moving items doesn't snap to the time divison selected.
Also there's a  bug making the time round upwards on the readouts. So if hh:mm:ss is selected, the seconds is updated 0.5 seconds too early. this is an old bug and it's woven tightly into how snapping works. I'm trying to solve that.

**Reply by akleja** [reply]
I'm trying to make all the changes as compatible as possible with the goal of eventually having a new default layout, but also an easy way of loading the classic layout as a preset.
